src/ParserRuleContext.ts

```diff
@@ -46,7 +46,26 @@ export class ParserRuleContext extends RuleContext {
 	 *  operation because we don't the need to track the details about
 	 *  how we parse this rule.
 	 */
-	children?: ParseTree[];
+
+
+	_children?: Array<ParseTree>;
+	static _emptyChildren = [];
+
+	/** Access children for read
+	 * Avoids creation of extra empty arrays by using a shared _emptyChildren array
+	 */
+	get children() : ReadonlyArray<ParseTree> {
+		let result = this._children;
+		return result ? result : ParserRuleContext._emptyChildren;
+	}
+
+	/** Access the children in a read/write way.
+	 * This will create the _children array if needed.
+	 */
+	get mutableChildren() {
+		if (!this._children) this._children = [];
+		return this._children as Array<ParseTree>;
+	}
```

Comment:
> However, wouldn't it be then already enough to assign the result of children to a const var to make it immutable?

No, in Typescript `const` means you can't change what object a symbol is referencing rather than making the referenced object as immutable.   (Comparing to C++, it's like the difference between `const char * str` vs. `char * const`;  typescript's const is like the later, which is rarely used as I remember.